### ISSUE TITLE
[PROPOSAL] Stop giving players the ability to screw-up their game state.

### ISSUE THREAD

**Issue by ShiningMinions** [sunflower-land/sunflower-land#936]
# TLDR;

All scary warning walls of texts / modals, stating "YOU MUST SYNC TO CHAIN" absolutely need to be killed. How about we simply don't give the player the ability to screw-up their game state in the first place? Instead, [detect if a player's game state is out of sync in the background](https://github.com/sunflower-land/sunflower-land/pull/855). When a player is about to make a state-breaking action, first prompt them with message: "You must sync on chain before performing this action". (Bonus points if a "sync now" button shows here as well).

![scary-modal-1](https://user-images.githubusercontent.com/103600068/168704199-34cfd3f9-54e0-4fa5-9733-a099ed9937e4.png)

![scary-modal-2](https://user-images.githubusercontent.com/103600068/168704378-51d12af4-b4d6-4965-b4b9-9d331dfbf4c4.png)

# The Problem

All over the game are scary warning messages / walls of texts warning the player: "YOU MUST SYNC TO CHAIN!!! AHHH!!!" before performing certain actions (otherwise the player risks losing serious progress). I'm thinking especially about the players who are new to the blockchain world. Texts like these are extremely intimidating, makes the player anxious of making a wrong-move, and is a horrible user experience.

# Proposed Solution

How about we simply DON'T allow the player to perform these state breaking on-chain actions? If they're about to perform one of these actions with an out-of-sync state, simply DISABLE the state-breaking button, and instead show them the text "You must sync on chain before performing this action". The player is then presented with a "sync now" button instead. Finally, the player clicks the "sync now" button, and performs traditional sync workflow. Finally, once sync confirms, the original action button is re-enabled, allowing them to perform their original intended action (i.e. withdraw, mint).

We already know we can calculate if a player's state is out of sync here: https://github.com/sunflower-land/sunflower-land/pull/855 , which directly supports this proposal's solution.

## Example Scenario

In this scenario, let's call our player "Bob" 🧑‍🌾. Bob has been planting and harvesting sunflowers for the past hour. Bob decides that he now wants to withdraw some wood he chopped yesterday (but oops! Bob went to Goblin Town and forgot to sync to chain first).

So now, Bob is in Goblin Town, and he opens the item withdraw modal at the goblin bank. Here is what Bob sees, both for how the game works now VS how it should work:

1. **HOW IT WORKS NOW** - Bob sees a wall of warning text (that he sees everyday) and has been conditioned to click through and not read it. He goes through the withdraw workflow like normal. Bob is now sad because all of his hard work from farming sunflowers is GONE 😭.
3. **HOW IT SHOULD WORK INSTEAD** - At no point is Bob ever presented with a wall of scary red warning text. Instead, when he goes to withdraw items at the bank, he notices that the "Withdraw" button is disabled. Above this disabled "Withdraw" button, he sees the following message: "You must sync to chain before performing this action", coupled with a "sync now" button. Bob happily clicks the "sync now" button. Once he goes through the sync workflow, the "Withdraw" button is now enabled, and Bob can start choosing the wood he wants to withdraw. Happy days! Bob gets to withdraw his wood AND keep his sunflowers 🤗.

**Comment by adamhannigan** [comment]
@ShiningMinions Great idea 👏 

From a game POV it is extremely important how we convey this to a user. Especially since we are blocking them from performing an action, we don't want them to think it is an issue with our tech, but rather part of the game.

Do you have any ideas how we could narrate this to the player or use some sort of gameplay/lore to explain why they need to sync?

**Comment by ShiningMinions** [comment]
@adamhannigan I will start thinking about how this concept can be narrated/"lore-ifed" to the player. I will write another comment when I have an idea. However, I have arguments why I think you can get away without needing to "lore-ify" this to the player.

> Especially since we are blocking them from performing an action, we don't want them to think it is an issue with our tech, but rather part of the game.

Here's my rebuttal - look at what happens when a player buys all the "seeds" in the shop. The "Buy" buttons go away, and they are presented with a "Sync" button. As a player, this is extremely clear to me what action I need to take next to continue. At no point do I think anything is "wrong with the game", it's just something I need to do to continue.

Can we not have something like this for minting/withdrawals too?

**Comment by sir-canada** [comment]
I learnt from dealing with lots of people in the discord that the biggest problem is 90% of the people playing this game are not native English speaking people. They're guessing at what prompts mean and act like everything is that 'user agreement' you click before you install or sign up for anything. 

The sync button is a really good idea.

Maybe a middle ground? Put a 'sync' button in the goblin village prompt. 'Hide' an 'advanced user' option in the menu settings, where reset session is. Saved as a cookie. If an user clicks the advanced user checkbox, it will allow him to go to goblin village without syncing.

This will allow for users to go to goblin village without syncing, but not without thinking about what they're doing first. If they're clicking the advanced user checkbox and lose items, that's on them.

**Comment by cambrios1** [comment]
@adamhannigan 

@ShiningMinions makes a great point--there are tons of questions concerns about this in discord. 

Lore/text on farm side:
Goblin Village is where you can craft rare NFTs, deposit your LP tokens in the Wishing Well, and withdraw SFL or items from the game. Feel free to look around, but make sure you sync your farm before transacting with any Goblins!

Goblin Village side functionality:
For an un-synced farm, make the sync button mandatory before they can click the craft button. Better yet, build the sync into the first transaction undertaken in Goblin Village.

**Comment by hujikol** [comment]
got one idea, 
by removing the red flag message, 
-> should be added a prompt message whenever action needs a 'sync on chain' 
-> and also showing last time sync 
-> and or also showing the difference between last time sync and current progress, (so at least the user knows what might be lost in the transaction if they are not 'sync on chain' before).

**Comment by ShiningMinions** [comment]
@hujikol thanks for your input!

> -> should be added a prompt message whenever action needs a 'sync on chain'

Indeed, this is part of my original proposal above ☝️.

> -> and also showing last time sync

This is a good idea, but I don't think this is relevant for fixing the proposed problem.

> -> and or also showing the difference between last time sync and current progress, (so at least the user knows what might be lost in the transaction if they are not 'sync on chain' before).

This I disagree with - why give the player the ability to screw up their game state? Nobody wants this - it needs to stop because it happens way too often accidentally. There is simply 0 use cases why any average player would want to screw up their game state when performing an on-chain action.

Even if some weird player REALLY wants to lose all of their progress since last sync, then there's a simple already existing solution - it's called the "Reset Session" button. The "Reset Session" button would be a better place for "showing the difference between last time sync and current progress, so the player knows what they will about to lose".

**Comment by adamhannigan** [comment]
Thanks for the contribution and feedback! ❤️ 

 We are moving towards using an ideas portal to help capture and vote on community ideas. Github will only be used for bug/technical reports. We are only a small team and it is becoming difficult for us to respond individually to every case so we have began using community voting software.   

 If you could please upload your idea to the following community portal it would be greatly appreciated 🤗   

 https://sunflower-land.ideas.aha.io/

 You can vote on similar ideas and help bring ideas to our attention. Thanks again for being apart of this community!
